### PR TITLE
cql3: expr: make column_value (and similar) a first-class expression

### PR DESCRIPTION
**Pull request by avikivity**
Currently, column names can only appear in a boolean binary expression,
but not on their own. This means that in the statement

   SELECT a FROM tab WHERE a > 3;

We can represent the WHERE clause as an expression, but not the selector.

To pave the way for using expressions in selector contexts, we promote
the elements of binary_operator::lhs (column_value, column_value_tuple,
token) to be expressions in their own right. binary_operator::lhs
becomes an expression (wrapped in unique_ptr, because variants can't
contain themselves).

Note that all three new possibilities make sense in a selector:

  SELECT column FROM tab
  SELECT token(pk) FROM tab
  SELECT function_that_accepts_a_tuple((col1, col2)) FROM tab

There is some fallout from this:

 - because binary_operator contains a unique_ptr, it is no longer
   copyable. We add a copy constructor and assignment operator to
   compensate.
 - often, the new elements don't make sense when evaluating a boolean
   expression, which is the only context we had before. We throw in
   these cases. The parser right now prevents such cases from being
   constructed in the first place (this is equivalent to if (some_struct_value)
   in C).
 - in statement_restrictions.cc, we need to evalute the lhs in the context
   of the full binary operator. I introduced with_current_binary_operator()
   for this; an alternative approach is to create a new sub-visitor.

### CODE REVIEW

File: cql3/expr/expression.hh [hunk 1]
```diff
@@ -95,10 +98,17 @@ enum class comparison_order : char {
 
 /// Operator restriction: LHS op RHS.
 struct binary_operator {
-    std::variant<column_value, column_value_tuple, token> lhs;
+    std::unique_ptr<expression> lhs;
     oper_t op;
     ::shared_ptr<term> rhs;
-    comparison_order order = comparison_order::cql;
+    comparison_order order;
```
Comment (nyh) [inline cql3/expr/expression.hh:104]: I'm really curious (i.e., would have been nice to have a comment) what is this "comparison order thing" and why it's part of a binary operator.

File: cql3/restrictions/statement_restrictions.cc [hunk 2]
```diff
@@ -145,27 +145,55 @@ to_column_definition(const schema_ptr& schema, const ::shared_ptr<column_identif
...
+        const binary_operator* current_binary_operator = nullptr;
 
         void operator()(const conjunction& c) {
             std::ranges::for_each(c.children, [this] (const expression& child) { std::visit(*this, child); });
         }
 
         void operator()(const binary_operator& b) {
-            if (std::holds_alternative<token>(b.lhs)) {
+            if (current_binary_operator) {
+                throw std::logic_error("Nested binary operators are not supported");
```
Comment (nyh) [inline cql3/restrictions/statement_restrictions.cc:177]: An special reason why? (or just that you didn't want to implement it yet)

**Review by nyh — COMMENTED** [review]
Looks good, but I'd like someone who has some context of the new expression code and where it's going (namely, @dekimir) to also review.
However, something I'm really missing in this patch is CQL tests (in one of the 17 frameworks we have for that) that demonstrate the new capabilities and all their corner cases and check any regressions in them. For example, from your example I didn't understand if "SELECT c1 + c2 FROM tab" will now work. If I saw actual examples in a test I would have understood.
Additionally, if we're adding features to CQL that don't exist in Cassandra, I think we should document somewhere exactly what those features are. We already have a place for such documentation: docs/design-notes/cql-extensions.md

File: cql3/restrictions/statement_restrictions.cc (hunk 2 above)
Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:177]: There's no way to create a nested binary expression from the parser, so even if I implemented it, it would never get exercised.

I'd like to generalize the parser in the future so that this is possible.

Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:177]: (i.e. I expect never to see this thrown).

File: cql3/expr/expression.hh (hunk 1 above)
Comment (dekimir) [reply cql3/expr/expression.hh:104]: It's literally 10 lines above...

**Review by dekimir — COMMENTED** [review]
Not done reviewing yet, just two minor comments...

File: cql3/expr/expression.hh (hunk 1 above)
Comment (nyh) [reply cql3/expr/expression.hh:104]: I didn't see it in the github presentation of the patch (I thought it was something that Avi added).
Ok, now I understand it but I can't say I like it very much: It sounds to me like there are two different "<" operators - the cql one, and the clustering one (by the way, I don't understand what the second one means... Does it mean the tokens are compared for partition key? or something else). So you can have two operators - LT_CQL and LT_CLUSTERING. The general flag you have there suggests it applies to *every* operator. But does it apply to "="? Does it apply to "+"? I guess the answer is no.

Comment (dekimir) [reply cql3/expr/expression.hh:104]: > I don't understand what the second one means

See the patch that introduced SCYLLA_CLUSTERING_BOUND, it explains the reasons.

> The general flag you have there suggests it applies to every operator

Every _ordering_ operator.

Comment (dekimir) [reply cql3/expr/expression.hh:104]: Here's a discussion that led to the introduction of SCYLLA_CLUSTERING_BOUND:
https://github.com/scylladb/scylla-tools-java/pull/224#discussion_r573760181

**Comment by avikivity** [comment]
> Looks good, but I'd like someone who has some context of the new expression code and where it's going (namely, @dekimir) to also review.
> However, something I'm really missing in this patch is CQL tests (in one of the 17 frameworks we have for that) that demonstrate the new capabilities

There are no new capabilities.

> and all their corner cases and check any regressions in them. For example, from your example I didn't understand if "SELECT c1 + c2 FROM tab" will now work. If I saw actual examples in a test I would have understood.

This is needed to make c1+c2 work, but doesn't now.

It's also for existing queries likes UPDATE tab SET my_counter = my_counter + 1 or UPDATE tab SET my_list = ? + my_list, but that will only be complete once selectors and update expressions are migrated to use expressions. There are now using Java-derived OO goo.

> Additionally, if we're adding features to CQL that don't exist in Cassandra, I think we should document somewhere exactly what those features are. We already have a place for such documentation: docs/design-notes/cql-extensions.md

This patch doesn't add features.

File: cql3/expr/expression.cc [hunk 3]
```diff
@@ -876,16 +939,15 @@ expression replace_column_def(const expression& expr, const column_definition* n
...
-                return std::visit(overloaded_functor{
-                        [&] (const column_value& col) {
-                            return expression(binary_operator{column_value{new_cdef}, oper.op, oper.rhs});
-                        },
-                        [&] (const column_value_tuple& tuple) -> expression {
-                            throw std::logic_error(format("replace_column_def invalid LHS: {}", to_string(oper)));
-                        },
-                        [&] (const token&) { return expr; },
-                    }, oper.lhs);
+                return expression(binary_operator(replace_column_def(*oper.lhs, new_cdef), oper.op, oper.rhs, oper.order));
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:942]: I did not intend to fix a bug here, and such bug fixes should not be hidden in large patches. I'll send a separate fix for this, and then rebase my patch on top.

[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:942]: Converted to draft pending this bugfix.

File: cql3/expr/expression.hh (hunk 1 above)
Comment (avikivity) [reply cql3/expr/expression.hh:104]: > > I don't understand what the second one means
> 
> See the patch that introduced SCYLLA_CLUSTERING_BOUND, it explains the reasons.
> 
> > The general flag you have there suggests it applies to every operator
> 
> Every _ordering_ operator.
> 
> > Does it apply to "+"? I guess the answer is no.
> 
> Correct.  That's because there is no `+`.

CQL has a +, so that's not the reason.

Comment (dekimir) [reply cql3/expr/expression.hh:104]: > CQL has a +

`binary_operator` doesn't.

File: cql3/expr/expression.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:942]: This is more involved. This function is used when converting a restriction on some column in the base table to a restriction on the materialized view that backs the index.

If the restriction is equality, then `order` doesn't affect anything.

If the restriction is ordering, then there are two cases:

1. it's a single-column restriction, in which case `order` was the default cql ordering and this doesn't change anything.
2. it's a multi-column restriction, in which case no index will be eligible and we won't reach this point.

So no bug is introduced. I'll drop the last parameter to make sure there is no unintended change.

Ultimately I think replace_column_def() is miguided. The caller should build a new restriction from scratch, not attempt to do a deep search and replace, so that it is sure about what it is building.

File: cql3/expr/expression.hh (hunk 1 above)
Comment (avikivity) [reply cql3/expr/expression.hh:104]: `binary_operator` surely will gain the ability to add/concatenate, or we won't be able to use expressions in update statements.

File: cql3/expr/expression.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc]: Updated with the change removed, it is now compatible with the previous code. We'll fix (or not) independently of this patch.

**Comment by avikivity** [comment]
Update: remove unintended change in replace_column_def(). The change should be considered separately. AFAICT the change could have no ill effect, but it's better not to sneak it in like that.

File: cql3/expr/expression.hh (hunk 1 above)
Comment (nyh) [reply cql3/expr/expression.hh:104]: @dekimir if this class is intended only for **comparison** operators, it should be called `comparison_operator`. If it is called `binary_operator` it suggests it will also support `+` et al. in the (not so far-away) future. And if it will support many different operators, adding the comparison order to all of them is really weird.
I still don't understand the "comparison order" thing needs to be part of the expression - and not knowledge in the **evaluator** of the expression (who knows if it's comparing a clustering column or not), or, alternatively, part of the operator itself (you'll have two versions of each of the operators <, <=, >, >= - but not for any other operator).

**Review by nyh — APPROVED** [review]
Looks good to me (I left a complaint about the "order" flag in the operator, but it's pre-existing code). I can commit this if @dekimir approves.

File: cql3/expr/expression.hh (hunk 1 above)
Comment (dekimir) [reply cql3/expr/expression.hh:104]: One day soon we'll separate out comparison operators into their own class with the `order` member, and this confusion will be avoided.  Or we will add different operators for different orders.  Until then, I feel it's acceptable to have this in `binary_operator`, even if it clearly doesn't apply to some operators.

> I still don't understand the "comparison order" thing needs to be part of the expression

It defines the meaning of the expression (see [here](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/test/cql/query_bounds_test.result#L23-55)).  Any reasonable use of the expression must know the order.

File: cql3/restrictions/statement_restrictions.cc [hunk 4]
```diff
@@ -174,7 +202,11 @@ static std::vector<expr::expression> extract_partition_range(
                         found->second = make_conjunction(std::move(found->second), b);
                     }
                 }
-            }
+            });
+        }
+
+        void operator()(const column_value_tuple& s) {
+            // Ignore tuples while evaluating single-column restrictions
```
Comment (dekimir) [inline cql3/restrictions/statement_restrictions.cc:209]: We're not limited to single-column restrictions here, but to partition-key restrictions (including token).  The real reason for ignoring tuples is that the parser disallows partition columns in tuples.

File: cql3/restrictions/statement_restrictions.cc [hunk 5]
```diff
@@ -897,6 +947,18 @@ struct multi_column_range_accumulator {
...
+    void operator()(const column_value&) {
+        throw std::logic_error("Column encountered outside binary operator");
+    }
+
+    void operator()(const column_value_tuple&) {
+        throw std::logic_error("Column tuple encountered outside binary operator");
+    }
+
+    void operator()(const token&) {
+        throw std::logic_error("Column encountered outside binary operator");
```
Comment (dekimir) [inline cql3/restrictions/statement_restrictions.cc:959]: "Token encountered"?

**Review by dekimir — CHANGES_REQUESTED** [review]
I like this direction, though I suggested some minor changes.  Flattening the visitors is great, I think, and with_current_binary_operator is perfectly acceptable.

File: cql3/restrictions/statement_restrictions.cc [hunk 6]
```diff
@@ -145,27 +145,55 @@ to_column_definition(const schema_ptr& schema, const ::shared_ptr<column_identif
 }
 #endif
 
+template <typename Visitor>
+concept visitor_with_binary_operator_context = requires (Visitor v) {
+    { v.current_binary_operator } -> std::same_as<const expr::binary_operator*>;
+};
+
+void with_current_binary_operator(
```
[the remark this answers is not in the document]
Comment (dekimir) [reply cql3/restrictions/statement_restrictions.cc:153]: On second thought, it's basically a two-liner, so it's fine without a blurb.

File: cql3/expr/expression.cc [hunk 7]
```diff
@@ -807,7 +863,16 @@ bool is_supported_by(const expression& expr, const secondary_index::index& idx)
                             return false;
                         },
                         [&] (const token&) { return false; },
-                    }, oper.lhs);
+                        [&] (const binary_operator&) -> bool {
+                            throw std::runtime_error("Nested binary operators are not supported");
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:867]: They're mostly a don't care thing since they aren't expected to happen. I'll convert to on_internal_error() to make it clear.

File: cql3/expr/expression.hh [hunk 8]
```diff
@@ -184,6 +194,9 @@ const binary_operator* find_atom(const expression& e, Fn f) {
                 }
                 return nullptr;
             },
+            [] (const column_value&) -> const binary_operator* { return nullptr; },
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:197]: Well, atom doesn't describe binary_operator either. Atoms (the greek ones at least) were supposed to be indivisible.

btw, the patch doesn't change find_atom's return value - it can only return binary_operators.

Perhaps it should be renamed to

```c++
template <subexpression Subexpression, std::regular_invocable<const S&> Func>
const Subexpression*
find_subexpression(const expression& e, Func func);
```

(where `subexpression` is a concept that is true for any type that can be an `expression`)

Which will allow it to find anything.

File: cql3/restrictions/statement_restrictions.cc (hunk 4 above)
Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:209]: Is it the parser or something lower? Either way I'll adjust.

File: cql3/restrictions/statement_restrictions.cc (hunk 5 above)
Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:959]: yes

File: cql3/expr/expression.hh (hunk 8 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply cql3/expr/expression.hh:197]: I like that renaming; it can be a follow-up.

File: cql3/expr/expression.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:867]: (we can also patch on_internal_error to report the source location)

File: cql3/expr/expression.cc [hunk 9]
```diff
@@ -552,6 +581,15 @@ bool is_satisfied_by(const expression& restr, const column_value_eval_bag& bag)
                 });
             },
             [&] (const binary_operator& opr) { return is_satisfied_by(opr, bag); },
+            [] (const column_value&) -> bool {
+                throw std::runtime_error("A column cannot serve as a restriction by itself");
+            },
```
Comment (avikivity) [inline cql3/expr/expression.cc:586]: btw, this can one day be lifted. If we have a column of type bool, SQL accepts `WHERE col`.

**Comment by avikivity** [comment]
v3:
 - convert throws to on_internal_error
 - disambiguate error messages
 - improve comment in extract_patition_range column_value_tuple case

**Review by dekimir — APPROVED** [review]
LGTM

**Comment by avikivity** [comment]
@nyh can you please re-review? @dekimir has approved the patch.

**Comment by avikivity** [comment]
@nyh ping, everything is ready

**Comment by avikivity** [comment]
@nyh ping

**Comment by avikivity** [comment]
oh, you just queued. unping.
